engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/StraightThroughProcessingLoopValidator.java

```diff
@@ -0,0 +1,222 @@
+                STRAIGHT_THROUGH_PROCESSING_ELEMENT_TYPES.contains(flowElement.getElementType())
+                    && !WHITELISTED_ELEMENT_TYPES.contains(flowElement.getElementType()))
+        .map(ExecutableFlowNode.class::cast)
+        .toList();
+  }
+
+  /**
+   * This method is the part that actually tries to find any loops. It takes a LinkedList in which
+   * it keeps track of elements that are part of the (potential) loop. It also takes an element
+   * which is used to detect if we are in a loop.
+   *
+   * <p>Upon entering this method there are 3 possibilities:
+   *
+   * <ul>
+   *   <li>This element is part of the potentialLoop list. This means that we have passed this
+   *       element before. As a result we can conclude that we are in a loop.
+   *   <li>The element is not part of the potentialLoop list. We haven't passed the element before
+   *       so we are not in a loop yet. However, the element is a straight-through processing
+   *       element. We must check the outgoing sequence flow of this element to see if it loops back
+   *       around to an element that is part of the potentialLoop list.
+   *   <li>We have not encountered this element before, nor is it a straight-through processing
+   *       element. We can conclude that this path is not part of a straight-through processing
+   *       loop.
+   * </ul>
+   *
+   * @param potentialLoop a list which keeps track of elements that are part of a potential loop
+   * @param element the element used to determine if we reached a loop, or are definitely not part
+   *     of a loop
+   * @return an Either of which the left contains a list of the straight-through processing loop.
+   */
+  private static Either<List<ExecutableFlowNode>, ?> checkForStraightThroughProcessingLoop(
+      final LinkedList<ExecutableFlowNode> potentialLoop, final ExecutableFlowNode element) {
+
+    if (foundLoop(potentialLoop, element)) {
+      // It could happen that we detect a loop which doesn't involve the first element we checked.
+      // By taking a sublist we make sure we've isolated the loop.
+      final var loop = potentialLoop.subList(potentialLoop.indexOf(element), potentialLoop.size());
+      // This element is added to the loop twice. This is useful when creating an error message
+      // describing the loop.
+      loop.add(element);
+      return Either.left(loop);
+    } else if (STRAIGHT_THROUGH_PROCESSING_ELEMENT_TYPES.contains(element.getElementType())) {
+      // We are not in a loop yet, but the element is a straight-through processing element. We must
+      // keep checking for loops by analysing the outgoing sequence flows of this element.
+      potentialLoop.addLast(element);
+      Either<List<ExecutableFlowNode>, ?> isPartOfLoop = Either.right(null);
+      for (final ExecutableSequenceFlow outgoing : element.getOutgoing()) {
+        isPartOfLoop = checkForStraightThroughProcessingLoop(potentialLoop, outgoing.getTarget());
```

Comment:
:x: Be aware recursion are always risky. To be specific this might be risky on large deployments right? Stack-overflow will fail the partition.

Reply:
@Zelldon You marked this as a blocker. Are you sure this fails the partition? I was under the impression that an unexpected error on a command would just result in a command rejection. If that's the case that would be fine for me. Users would have to model some very crazy processes to run into a SO exception. I tried it with a chain of 500 undefined tasks and didn't run into it.

Reply:
Throwable will not be catched by the StreamPlatform. https://github.com/camunda/zeebe/blob/main/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingStateMachine.java#L306

It will cause to fail the actor. 

Yeah I guess you might be right and the anxiety might be too excessive. I just try to be on the safe side in such cases, since we already have seen that several times.

https://github.com/camunda/zeebe/issues/9545 
https://github.com/camunda/zeebe/issues/7758

Reply:
I can turn it into an expected exception and reject it in the processor. In the rare case that it does happen it would at least prevent the partition from breaking.

Reply:
Would be a compromise for me :+1: